astroquery/heasarc/core.py

```diff
@@ -135,290 +455,293 @@ def query_object_async(self, object_name, mission, *,
-                           **kwargs):
+        Parameters
+        ----------
+        query_result : `astropy.table.Table`, optional
+            A table that contain the search resutls. Typically as
+            returned by query_region. If None, use the table from the
+            most recent query_region call.
+        tablename : str
+            The table name for the which the query_result belongs to.
+            If None, use the one from the most recent query_region call.
+
+        Returns
+        -------
+        table : A `~astropy.table.Table` object.
         """
-        Query around specific set of coordinates within a given mission
-        catalog. Method first converts the supplied coordinates into the FK5
-        reference frame and searches for sources from there. Because of this,
-        returned offset coordinates may look different than the ones supplied.
+        if query_result is None:
+            if (
+                not hasattr(self, '_query_result')
+                or self._query_result is None
+            ):
+                raise ValueError('query_result is None, and none '
+                                 'found from a previous search')
+            else:
+                query_result = self._query_result
+
+        if not isinstance(query_result, Table):
+            raise TypeError('query_result need to be an astropy.table.Table')
+
+        # make sure we have a column __row
+        if '__row' not in query_result.colnames:
+            raise ValueError('No __row column found in query_result. '
+                             'query_result needs to be the output of '
+                             'query_region or a subset.')
+
+        if tablename is None:
+            tablename = self._tablename
+        if not (
+            isinstance(tablename, str)
+            and tablename in self.tap.tables.keys()
+        ):
+            raise ValueError(f'Unknown table name: {tablename}')
+
+        # datalink url
+        dlink_url = f'{self.VO_URL}/datalink/{tablename}'
+
+        query = pyvo.dal.adhoc.DatalinkQuery(
+            baseurl=dlink_url,
+            id=query_result['__row'],
+            session=self._session
+        )
+        dl_result = query.execute().to_table()
+        dl_result = dl_result[dl_result['content_type'] == 'directory']
+        dl_result = dl_result[['ID', 'access_url', 'content_length']]
+
+        # add sciserver and s3 columns
+        newcol = [
+            f"/FTP/{row.split('FTP/')[1]}".replace('//', '/')
+            if 'FTP' in row else ''
+            for row in dl_result['access_url']
+        ]
+        dl_result.add_column(newcol, name='sciserver', index=2)
+        newcol = [f"s3://{self.S3_BUCKET}/{row[5:]}" if row != '' else ''
+                  for row in dl_result['sciserver']]
+        dl_result.add_column(newcol, name='aws', index=3)
+
+        return dl_result
+
+    def enable_cloud(self, provider='aws', profile=None):
```

Comment:
hopefully everthing cloud related will make its way up to pyvo, eventually, but until then it all can stay of course.